packages/peregrine/lib/talons/SignIn/useSignIn.js

```diff
@@ -47,48 +56,59 @@ export const useSignIn = props => {
         async ({ email, password }) => {
             setIsSigningIn(true);
             try {
-                // Sign in and save the token
-                const response = await signIn({
+                // Get source cart id (guest cart id).
+                const sourceCartId = cartId;
+
+                // Sign in and set the token.
+                const signInResponse = await signIn({
                     variables: { email, password }
                 });
-
-                const token =
-                    response && response.data.generateCustomerToken.token;
-
+                const token = signInResponse.data.generateCustomerToken.token;
                 await setToken(token);
-                await getUserDetails({ fetchUserDetails });
-
-                // Then remove the old guest cart and get the cart id from gql.
-                // TODO: This logic may be replacable with mergeCart in 2.3.4
-                await removeCart();
 
+                // Clear all cart/customer data from cache and redux.
                 await clearCartDataFromCache(apolloClient);
                 await clearCustomerDataFromCache(apolloClient);
+                await removeCart();
 
+                // Create and get the customer's cart id.
                 await createCart({
                     fetchCartId
                 });
+                const destinationCartId = await retrieveCartId();
+
+                // Merge the guest cart into the customer cart.
+                await mergeCarts({
+                    variables: {
+                        destinationCartId,
+                        sourceCartId
+                    }
+                });
 
-                await getCartDetails({ fetchCartId, fetchCartDetails });
+                // Ensure old stores are updated with any new data.
+                getUserDetails({ fetchUserDetails });
+                getCartDetails({ fetchCartId, fetchCartDetails });
```

Comment:
We've got this same web of code in a few different places now.

That said, I don't think it's worth refactoring because the talons are so specific. Just pointing out that it'll be easy for these to get out of sync and bugs to creep in.

Reply:
Not worth refactoring - we will be removing a lot of this when we do the work on new mini cart.